app/forms/mainstream_browse_page_archival_form.rb

```diff
@@ -0,0 +1,46 @@
+class MainstreamBrowsePageArchivalForm
+  include ActiveModel::Model
+  attr_accessor :tag, :successor, :successor_path
+
+  validates :successor_path, presence: true, valid_govuk_path: true, if: :redirecting_to_path?
+
+  def browse_pages
+    published_browse_pages - [tag]
+  end
+
+  def archive_or_remove
+    return false unless valid?
+
+    if tag.published?
+      TagArchiver.new(tag, successor_object).archive
+    else
+      DraftTagRemover.new(tag).remove
+    end
+
+    true
+  rescue GdsApi::HTTPConflict
+    errors.add :base, "The tag could not be deleted because there are documents tagged to it."
+    false
+  rescue GdsApi::HTTPClientError
+    errors.add :base, "The tag could not be deleted because of an error."
+    false
+  end
+
+private
+
+  def successor_object
+    if redirecting_to_path?
+      OpenStruct.new(base_path: successor_path, subroutes: [])
```

Comment:
Let's replace this with a Struct. You'll get a proper NoMethodError if a client object mishandles method calls on the successor_object, rather than the `nil` that OpenStruct returns.